plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSource.java

```diff
@@ -25,32 +26,84 @@
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.net.URI;
+import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 
+import static io.airlift.slice.Slices.EMPTY_SLICE;
+import static java.lang.Math.min;
 import static java.util.Objects.requireNonNull;
 
 public class FileSystemExchangeSource
         implements ExchangeSource
 {
-    private final FileSystemExchangeStorage exchangeStorage;
     @GuardedBy("this")
-    private final Iterator<URI> files;
+    private final List<CompletableFuture<Void>> readFutures;
     @GuardedBy("this")
-    private final Iterator<Optional<SecretKey>> secretKeys;
-
-    @GuardedBy("this")
-    private SliceInput sliceInput;
+    private final BlockingQueue<Slice> pendingSlices;
     @GuardedBy("this")
     private boolean closed;
 
-    public FileSystemExchangeSource(FileSystemExchangeStorage exchangeStorage, List<URI> files, List<Optional<SecretKey>> secretKeys)
+    private final AtomicInteger finishedFutures = new AtomicInteger();
+    private final AtomicReference<Throwable> failure = new AtomicReference<>();
+
+    public FileSystemExchangeSource(
+            FileSystemExchangeStorage exchangeStorage,
+            List<URI> files,
+            List<Optional<SecretKey>> secretKeys,
+            ExecutorService parallelReadExecutor,
+            int parallelReadMaxConcurrency,
+            int parallelReadQueueSize)
     {
-        this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
-        this.files = ImmutableList.copyOf(requireNonNull(files, "files is null")).iterator();
-        this.secretKeys = ImmutableList.copyOf(requireNonNull(secretKeys, "secretKeys is null")).stream().iterator();
+        requireNonNull(exchangeStorage, "exchangeStorage is null");
+
+        final int parallelReadConcurrency = min(files.size(), parallelReadMaxConcurrency);
+        final int partitionSize = (files.size() + parallelReadConcurrency - 1) / parallelReadConcurrency;
+        this.readFutures = new ArrayList<>(parallelReadConcurrency);
+        this.pendingSlices = new LinkedBlockingQueue<>(parallelReadQueueSize);
+        List<List<URI>> fileLists = Lists.partition(ImmutableList.copyOf(requireNonNull(files, "files is null")), partitionSize);
```

Comment:
Ideally it is better to create a queue of files and have each reader to pull one file at a time from the queue (to decrease an impact of skewed files / slow reads)

Reply:
This is a good one. As discussed, the "parallel read" commit is more like a proof-of-concept commit, and will not be included in this PR at the end. Will address this one separately.